### PR TITLE
refactor(protected-regions): replace hand-rolled ReDoS scanner with community analyzers

### PR DESCRIPTION
**Pull request by Jamie-BitFlight**
## Summary

Started as a fix for two round-13 ReDoS bypasses in the hand-rolled `screenExtraPatterns` complexity scanner (overlapping character ranges, a backreference losing capture context across an isolated group slice — both confirmed 4.5–4.7s hangs on 40-character adversarial input).

While closing those, the review cycle kept surfacing the same class of bug in the hand-rolled scanner — UTF-16/astral-character handling, capture-group renumbering across sliced analysis — each fixed and re-broken by the next edge case. That converged on replacing the entire hand-rolled `complexityRejection`/`atomCharSet`/`analyzeEmptiness` scanner with `analyseRegexSafety` (`src/core/regex-safety.ts`), which delegates to:

- [`recheck`](https://github.com/makenowjust-labs/recheck) for ReDoS/catastrophic-backtracking detection (automaton-based static analysis, with a working pure-JS fallback verified independently — see thread below)
- [`@eslint-community/regexpp`](https://github.com/eslint-community/regexpp) for AST parsing
- [`regexp-ast-analysis`](https://github.com/RunDevelopment/regexp-ast-analysis) for zero-length-match detection

`screenExtraPatterns` in `src/core/protected-regions.ts` is now a thin wrapper: length check, syntax check, then `analyseRegexSafety` for complexity and emptiness.

## Verification

- `vp check` — clean (format, lint, types)
- `vp pack` — clean build
- `scripts/ci/check-exit-codes.sh` — all five cases correct
- Full test suite: 637/637 passing (`test/unit/protected-patterns.test.ts` rewritten around the new interface — vulnerability, acceptance, and empty-match cases per Codex-reported and originally-authored edge cases)
- `docs/configuration.md` reconciled to describe the current screening behavior
- Independent review: Codex reviewed every commit through the current head (`83e43db`); all findings addressed or resolved through verified discussion (including one finding — a claimed missing ReDoS backend on platforms without a native module or JVM — that was reproduced as a false alarm: `recheck` 4.5.0 has a working pure-JS fallback, confirmed by forcing both external backend paths to nonexistent files)

## Test plan

- [x] `vp check`
- [x] Full test suite (637/637)
- [x] `vp pack`
- [x] `scripts/ci/check-exit-codes.sh`
- [x] Every Codex-reported regression pattern re-measured against Node's own regex engine, not just the screen's verdict

### CODE REVIEW

File: src/core/protected-regions.ts [hunk 1]
```diff
@@ -713,12 +765,28 @@ function atomCharSet(atomText: string): ReadonlySet<string> | undefined {
...
-    if (inner[k] === '-') return undefined;
+  const chars = new Set<string>();
+  let k = 0;
+  while (k < inner.length) {
+    // A `-` between two other characters is a range (`a-z`); only at either end, or alone, is it
+    // unambiguously the literal hyphen — the same reasoning `normalizeAtomText` already applies.
+    const isRange = inner[k + 1] === '-' && k + 2 < inner.length;
+    if (!isRange) {
+      chars.add(inner[k]!);
+      k += 1;
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Confirmed — reproduced exactly: `[😀]*[😁]*X` (two disjoint classes) was wrongly rejected as `adjacent-repetition`. Fixed in 826a904 by reading each class member via `codePointAt`/`fromCodePoint` instead of indexing one UTF-16 code unit at a time, so a supplementary-plane character is added as one whole set member regardless of whether it's a bare literal or (already correctly handled since round 13) a range endpoint.

Verified against the actual function: the reported pattern stays accepted; the same astral character repeated in two adjacent classes (`[😀]*[😀]*X`) and a mixed literal+astral class sharing one still correctly reject as overlapping — the fix corrects the false positive without weakening genuine overlap detection.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 2]
```diff
@@ -464,10 +472,28 @@ function canOnlyMatchEmpty(source: string): boolean {
...
+ * tracked for consumption — is conservatively treated as consuming.
+ *
+ * `outerGroupEmptyOnly`, when given, is consulted for a backreference key this walk's own
+ * `groupEmptyOnly` has no entry for. Needed because `source` is not always the whole pattern:
+ * {@link complexityRejection}'s `)` handler calls this on an isolated group body sliced out of the
+ * pattern, to decide whether that group can be treated as zero-width — and a backreference inside
+ * that slice can point at a group opened and closed entirely *outside* it, which this walk's own
+ * `collectTrackableGroupKeys(source)` (scoped to the slice) can never see. Found in external review
+ * of PR #73, round 13: `^()a*(?:\1)a*(?:\1)a*(?:\1)a*(?:\1)a*(?:\1)a*(?:\1)a*(?:\1)a*b$` confirmed
+ * 4.697s for 40 `a`s, because `canOnlyMatchEmpty('\1')` — called on the wrapper group's isolated
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Fixed in 826a904 — reworded both comments (this one and the sibling `4.537s` one) to describe the attack shape and point at the reproducible timing assertion in `test/unit/protected-patterns.test.ts` instead of recording a specific duration. Also caught and fixed a mislabeled "Round 12" in the sibling comment that should have said "Round 13" while I was in there.

Scoped to only the two comments this PR (round 13) itself introduced — there are ~30 similar specific-timing mentions in comments from rounds 1–12, already merged in #73. Sweeping those is a separate, larger cleanup this PR isn't taking on.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 3]
```diff
@@ -713,12 +771,37 @@ function atomCharSet(atomText: string): ReadonlySet<string> | undefined {
...
+      // surrogate pair — not two separate characters. Adding `inner[k]` alone would add just the
+      // high surrogate half as its own set member, so two *different* astral characters that
+      // happen to share the same high surrogate (common for emoji, which cluster in a handful of
+      // high-surrogate blocks) would appear to overlap on that shared half alone. Found in
+      // external review of PR #73/#76: `[😀]*[😁]*X` — two disjoint single-character classes —
+      // was wrongly rejected as `adjacent-repetition`. `codePointAt`/`fromCodePoint` read and add
+      // the whole character as one set member either way.
+      const codePoint = inner.codePointAt(k)!;
+      chars.add(String.fromCodePoint(codePoint));
+      k += codePoint > 0xffff ? 2 : 1;
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Confirmed — 5.1s for 40 `😀`s on this machine, accepted by the screen before fixing. Fixed in bd2f61f, and it went one layer deeper than the scanning-loop fix you suggested: advancing the bare-atom loop by codepoint alone wasn't sufficient, because `normalizeAtomText` and `atomCharSet`'s own single-atom branches both hardcoded `.length === 1`/`=== 3` checks that assume a bare or class-reducible atom is exactly one JS string-length unit — true for BMP, false for a two-unit surrogate pair. So even with the loop fixed, a bare astral atom's character set stayed `undefined` and its class form (`[😀]`) never reduced to the bare spelling, meaning neither the sameText nor the charSet-overlap comparison path ever actually ran between the two spellings. All three needed fixing together for `[😀]*😀*` to compare equal to `😀*😀*`.

Also found and fixed the identical one-code-unit-at-a-time bug in `analyzeEmptiness`'s sibling bare-atom path (same root cause, not part of your two reports): `😀{0}` was accepted instead of correctly classified `matches-only-empty`.

Verified against the actual function: the reported pattern rejected; disjoint/same-astral-character class-only cases from round 14 still correct; full prior-round corpus unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 4]
```diff
@@ -713,12 +771,37 @@ function atomCharSet(atomText: string): ReadonlySet<string> | undefined {
...
-  // A `-` between two other characters is a range (`a-z`); only at either end, or alone, is it
-  // unambiguously the literal hyphen — the same reasoning `normalizeAtomText` already applies.
-  for (let k = 1; k < inner.length - 1; k += 1) {
-    if (inner[k] === '-') return undefined;
+  const chars = new Set<string>();
+  let k = 0;
+  while (k < inner.length) {
+    // A `-` between two other characters is a range (`a-z`); only at either end, or alone, is it
+    // unambiguously the literal hyphen — the same reasoning `normalizeAtomText` already applies.
+    const isRange = inner[k + 1] === '-' && k + 2 < inner.length;
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Confirmed — `[😀-😁]*-*X` produced a 3-member set (both emoji plus a spurious literal `-`) instead of correctly recognizing a range, exactly as described. Fixed in bd2f61f: the range check now computes the starting codepoint's length first and looks for `-` at that offset, not a fixed one-code-unit offset, then bails (leaves the range unenumerated) if either endpoint is astral — same conservative default an oversized BMP range already gets.

Verified against the actual function: the reported pattern now stays accepted; ordinary BMP ranges (`[a-z]`, `[b-z]`) and the round-13/14 overlap cases are unaffected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 5]
```diff
@@ -690,20 +743,49 @@ function normalizeGroupBody(bodyText: string): string {
...
- * `[a-z]`, `[^a]`, …): not because those provably can't overlap with another atom, but because
- * this parser does not attempt to prove it, the same "provable, not exhaustive" bias as the rest
- * of this screen. Used to catch two adjacent range-quantified atoms that are not textually
- * identical but can still match the same character — found in external review of PR #73
- * (`a*[ab]*a*[ab]*a*[ab]*a*[ab]*b`, confirmed 5.066s for 40 `a`s before this fix, because the
- * single-character-class normalization above only recognises `[a]` as equivalent to `a`, not `[ab]`
- * as *overlapping* with `a`).
+ * a bare literal character, or a character class made up entirely of individual literal characters
+ * and/or literal ranges (`a-z`) below {@link MAX_EXPANDED_CLASS_SIZE} combined members, with no
+ * escape or negation. `undefined` for anything else (`.`, `\d`, `[^a]`, an oversized range, …): not
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Confirmed — this overview comment above `complexityRejection` still described the pre-round-13 behavior, contradicting both `atomCharSet`'s own doc comment and `docs/configuration.md`. Fixed in 0f56a80: reworded to say `[a-z]+` next to `[0-9]+` is now accepted because it's proven disjoint, not because it went unanalysed, matching what's actually implemented.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 6]
```diff
@@ -713,12 +795,51 @@ function atomCharSet(atomText: string): ReadonlySet<string> | undefined {
...
+      // external review of PR #73/#76, round 15: the *previous* code-unit-relative range check
+      // misread an astral start character's own low surrogate as the `-`, producing a 3-member set
+      // (both emoji plus a spurious literal `-`) for `[😀-😁]` instead of correctly recognising a
+      // range and bailing on its astral endpoints — `[😀-😁]*-*X` was wrongly rejected as
+      // overlapping, since the miscomputed set really did contain `-`.
+      if (startCodePoint > 0xffff || endCodePoint > 0xffff || endCodePoint < startCodePoint) {
+        return undefined;
+      }
+      if (chars.size + (endCodePoint - startCodePoint + 1) > MAX_EXPANDED_CLASS_SIZE) {
+        return undefined;
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Investigating — will confirm and fix shortly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/protected-regions.ts [hunk 7]
```diff
@@ -518,10 +547,13 @@ function analyzeEmptiness(source: string): {
...
-        // walk hasn't reached yet — always empty, see `collectTrackableGroupKeys` — or a
-        // reference to a group permanently out of scope, conservatively treated as consuming.
-        const resolved = groupEmptyOnly.get(key);
+        // order source appears in) — checked first against this walk's own map, then against the
+        // caller-supplied full-pattern context for a group this isolated slice never saw open or
+        // close at all (see the `outerGroupEmptyOnly` doc above). Neither resolving means either a
+        // forward reference to a group this walk hasn't reached yet — always empty, see
+        // `collectTrackableGroupKeys` — or a reference to a group permanently out of scope,
+        // conservatively treated as consuming.
+        const resolved = groupEmptyOnly.get(key) ?? outerGroupEmptyOnly?.get(key);
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/protected-regions.ts]: Investigating — will confirm and fix shortly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/core/regex-safety.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Apply community-owned ECMAScript parsing, zero-length analysis, and ReDoS detection.
+ */
+export function analyseRegexSafety(source: string): RegexSafetyResult {
+  try {
+    const literal = parseRegExpLiteral(new RegExp(source, 'gu'));
+    const matchesOnlyEmpty = isZeroLength(literal.pattern.alternatives, literal.flags);
+    const diagnostics = checkSync(source, 'gu', {
+      checker: 'auto',
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/regex-safety.ts:18]: Confirmed — verified against `recheck`'s own resolution logic in `node_modules/recheck/lib/main.js`:

- `recheck`'s `optionalDependencies` (checked via `npm view recheck-linux-arm64 versions`) publish native backends only for `linux-x64`, `macos-arm64`, `macos-x64`, `windows-x64` at the pinned `4.5.0` — `recheck-linux-arm64` only exists as `4.6.0-beta.3`/`4.6.0-beta.4`, no stable release.
- On an unmatched platform/arch, `recheck` falls back to `java -jar recheck.jar agent` (confirmed in the bundled resolver: it spawns `java` against the `recheck-jar` optional package) — there is no pure-JS/WASM path for either the `automaton` or `fuzz` checker.
- `java` is not declared anywhere as a system requirement — not in `engines`, not in the README, not checked at install/build time.
- `npm ci --omit=optional` drops `recheck-jar` too, so that install mode fails on every platform, not just Linux ARM64.

The failure mode is real: on Linux ARM64 without a JVM, `checkSync` throws, the catch in `analyseRegexSafety` turns that into `status: 'inconclusive'`, and `screenExtraPatterns` rejects every `extraProtectedPatterns` entry outright — the feature goes silently non-functional rather than failing loudly.

I'm not fixing this unilaterally: which direction to take is a dependency/architecture call (require + verify Java as a documented system dependency on non-native platforms, add a non-native pure-JS fallback checker, or treat "no backend available" as a distinct startup-time error instead of a per-pattern `inconclusive`), and this detection strategy has changed twice today. Let me know which direction you want and I'll implement it — happy to take a first pass at any of the three if you'd rather I just pick one.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/regex-safety.ts:18]: The reported backend gap does not occur in recheck 4.5.0. Its bundled `lib/main.js` contains a pure JavaScript implementation. The default synchronous path is `checkSync -> synckit worker -> check(auto)`; if both native and Java agents fail, `check(auto)` creates its bundled worker pool and executes that pure implementation.

I verified the fallback directly by forcing both external backend paths to nonexistent files:

```
RECHECK_BIN=/definitely-not-present/recheck \
RECHECK_JAR=/definitely-not-present/recheck.jar node ...
```

With the normal `checkSync` call, `PN\\d+` returned `safe` and `(a+)+$` returned `vulnerable` with exponential complexity. I also forced `RECHECK_SYNC_BACKEND=pure` separately and received the same exponential verdict. Neither test used a native backend or usable JAR.

The optional native/JAR packages accelerate analysis; they are not required for functionality. No project fallback or platform restriction is needed.

[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply src/core/regex-safety.ts:18]: Reproduced independently — same result. Ran your exact repro (`RECHECK_BIN`/`RECHECK_JAR` pointed at nonexistent paths) against `checkSync` directly:

```
safe test:  {"checker":"automaton","complexity":{"type":"linear",...},"status":"safe"}      // PN\d+
vuln test:  {"checker":"automaton","complexity":{"type":"exponential",...},"status":"vulnerable"}  // (a+)+$
```

Both correct, no native binary or JVM reachable, only a stderr line from the failed JAR-agent attempt (not a thrown error). Confirms your finding — the JS-side fallback inside `synckit`'s worker pool handles it once both external agents fail to spawn, which I hadn't traced past the resolver call in `lib/main.js`.

My previous reply overstated the risk — the platform gap I described doesn't actually cause `analysis-inconclusive`/feature loss the way I claimed. No code change needed here. Thanks for the correction.

---
_Generated by [Claude Code](https://claude.ai/code)_
